nomad/job_endpoint.go

```diff
@@ -58,8 +58,8 @@ func NewJobEndpoints(s *Server) *Job {
 		srv:    s,
 		logger: s.logger.Named("job"),
 		mutators: []jobMutator{
-			jobConnectHook{},
 			jobCanonicalizer{},
+			jobConnectHook{},
```

Comment:
Note to reviewers: this is one of those fun things that looks super minor but is actually the key to the entire PR.

Previously we were canonicalizing *after* adding Connect tasks.
This change makes it so we canonicalize *before* adding Connect tasks.

This solves the interpolation issue linked from the PR, but is also generally safer as just about everywhere in `nomad/` we touch Job structs we expect them to be canonicalized. So this should make future work in mutators much more predictable.

The downside is that any mutations must canonicalize themselves. The conservative version of this approach would just canonicalize twice: once at the beginning and once at the end. Since there is so little mutator code today I went with this minimal approach where the connect hook canonicalizes anything it adds.